app/controllers/admin/institutions_controller.rb

```diff
@@ -1,64 +1,69 @@
+    @institution = Institution.new
   end
 
-  # POST /admin/action_pages/:action_page_id/institutions
   def create
-    @institution = Institution.find_or_initialize_by(name: institution_params[:name])
-    @actionPage.institutions << @institution
-
-    respond_to do |format|
-      if @institution.save
-        format.html { redirect_to [:admin, @actionPage, Institution], notice: "Institution was successfully created." }
-      else
-        format.html { render "new" }
-      end
+    @institution = Institution.find_or_initialize_by(institution_params)
+
+    if @institution.save
+      redirect_to "index", notice: "#{@institution.name} was successfully created."
+    else
+      render "index"
     end
   end
 
-  # POST /admin/action_pages/:action_page_id/institutions/import
-  def import
-    names = []
-    CSV.foreach(params[:file].path, headers: true) do |row|
-      params = row.to_hash
-      unless params["name"]
-        flash[:notice] = "Import failed. Please check CSV formatting"
-        break
-      end
-      names << params["name"]
+  def edit; end
+
+  def update
+    if @institution.update(institution_params)
+      redirect_to "index", notice: "#{@institution.name} was successfully updated."
+    else
+      render "edit"
     end
+  end
 
-    Institution.delay.import(names, @actionPage) if names.present?
+  def upload; end
 
-    redirect_to [:admin, @actionPage, Institution]
+  def import
+    names = Institution.process_csv(params[:file])
+    if names.empty?
+      redirect_to "upload", notice: "Import failed. Please check CSV formatting"
+    else
+      Institution.delay.import(import_params[:category], names)
+      redirect_to "index", notice: "Successfully imported #{names.length} targets"
+    end
   end
 
-  # DELETE /admin/action_pages/:action_page_id/institutions/1/
   def destroy
-    @actionPage.institutions.delete(@institution)
-    respond_to do |format|
-      format.html { redirect_to [:admin, @actionPage, Institution] }
-    end
+    @institution.destroy
+    redirect_to "index", notice: "Target deleted"
   end
 
-  # DELETE /admin/action_pages/:action_page_id/institutions/
-  def destroy_all
-    @actionPage.institutions.delete_all
-    respond_to do |format|
-      format.html { redirect_to [:admin, @actionPage, Institution] }
-    end
+  def edit_targets
+    @institutions = Institution.all.order(:name)
+    @action_institutions = @actionPage.institutions.order(:name)
+    @not_included_institutions = @institutions - @action_institutions
+    @grouped_institutions = @institutions.group_by(&:category)
+    5.times { @actionPage.affiliation_types.build }
+  end
+
+  def update_targets
+    @actionPage.update(affiliation_types_params)
+    ActionInstitution.update_action_page(action_page: @actionPage,
+                                         add: institutions_diff_params[:add_ids],
+                                         remove: institutions_diff_params[:remove_ids])
```

Comment:
possibly should do this in the background like importing institutions, it takes a while